ctapipe/reco/shower_processor.py

```diff
@@ -0,0 +1,193 @@
+"""
+High level processing of showers.
+
+This processor will be able to process a shower/event in 3 steps:
+- shower geometry
+- estimation of energy (optional, currently unavailable)
+- estimation of classification (optional, currently unavailable)
+
+"""
+import numpy as np
+
+from astropy.coordinates import SkyCoord, AltAz
+
+from ctapipe.core import Component, QualityQuery
+from ctapipe.core.traits import List
+from ctapipe.containers import ArrayEventContainer, ReconstructedShowerContainer
+from ctapipe.instrument import SubarrayDescription
+from ctapipe.reco import HillasReconstructor
+
+
+DEFAULT_SHOWER_PARAMETERS = ReconstructedShowerContainer()
+
+
+class ShowerQualityQuery(QualityQuery):
+    """Configuring shower-wise data checks."""
+
+    quality_criteria = List(
+        default_value=[
+            ("Positive charge", "lambda hillas: hillas.intensity > 0"),
+            ("Positive width", "lambda hillas: hillas.width.value > 0"),
+            ("Positive length", "lambda hillas: hillas.length.value > 0"),
+            ("Ellipticity > 0.1", "lambda hillas: hillas.width.value / hillas.length.value > 0.1"),
+            ("Ellipticity < 0.6", "lambda hillas: hillas.width.value / hillas.length.value < 0.6")
+        ],
+        help=QualityQuery.quality_criteria.help,
+    ).tag(config=True)
+
+
+class ShowerProcessor(Component):
+    """
+    Takes DL1/parameters data and estimates the shower geometry.
+    Should be run after ImageProcessor to produce all DL1b information.
+
+    For now the only supported reconstructor is HillasReconstructor.
+    """
+
+    def __init__(
+        self,
+        subarray: SubarrayDescription,
+        is_simulation,
+        reconstruct_energy,
+        classify,
+        config=None,
+        parent=None,
+        **kwargs,
+    ):
+        """
+        Parameters
+        ----------
+        subarray: SubarrayDescription
+            Description of the subarray. Provides information about the
+            camera which are useful in calibration. Also required for
+            configuring the TelescopeParameter traitlets.
+        is_simulation: bool
+            If true, also process simulated images if they exist
+        config: traitlets.loader.Config
+            Configuration specified by config file or cmdline arguments.
+            Used to set traitlet values.
+            This is mutually exclusive with passing a ``parent``.
+        parent: ctapipe.core.Component or ctapipe.core.Tool
+            Parent of this component in the configuration hierarchy,
+            this is mutually exclusive with passing ``config``
+        """
+
+        super().__init__(config=config, parent=parent, **kwargs)
+        self.subarray = subarray
+        self.check_shower = ShowerQualityQuery(parent=self)
+        self._is_simulation = is_simulation
+        self.reconstruct_energy = reconstruct_energy
+        self.classify = classify
+        self.reconstructor = HillasReconstructor()
```

Comment:
Add a telescope parameter traitlet `reconstructor_type` and use `from_name`. Like in the `CameraCalibrator` for the extractors.

Reply:
If I do this I have to remove `Reconstructor` as a parent class from `Impact` - this is what I meant in the description of the PR.

Reply:
Why?

Reply:
because ImPACT as far as I know needs HillasReconstructor to work

in fact  ImPACT-like classes constitute (or should constitute) a different reconstruction class by themselves, for sure more than what Reconstructor does know (which to me now seems at least not up-to-date with what we have in mind). 

Generalizing Reconstructor to initialize all these choices (meaning HillasReconstructor and ImPACT for example) seems too complicated than making ImPACT not rely on Reconstructor at all

especially because after merging #1468 ImPACT will be really different in how it operates

it might work (meaning not crash with defaults) in master now, but as a choice further down the line I'd say it's better to treat it differently

Reply:
perhaps @kosack can integrate what I said if I missed something from what we discussed

Reply:
Yes, we discussed this off-line: the current way things are set up this will not work, except for HillasReconsructor and HillasIntersection (which both inherit from Reconstructor and do actually share a common  predict function), ImPACT works differently, but still inherits from Reconstructor.  It can be considered an "advanced reconstructor" which would be called in a *second step* following the standard reconstructor (since it needs the first-pass reconstruction as input), so it does not make sense to have it an option here.  However, the other reconstructors could be, so just removing the base class Reconstructor from ImPACT would fix the problem (and it doesn't use anything from that base class, so it should not break anything).  But re-factoring the existing reconstructions is out of scope of this PR, so should be done separately.   Once that is done, supporting an option to choose the type should be fine.

Reply:
Ok, then I would just (for now) remove the base class from `ImPACT`.

Reply:
Even if this doesn't really change anything here, I'd do this change separately

Reply:
ok, so for now, it's always `HillasReconstructor` and we can make another PR for enabling the choice of reconstructor.